pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1189,65 +1190,70 @@ protected void handleProducer(final CommandProducer cmdProducer) {
-                        } else if (retentionPolicy == BacklogQuota.RetentionPolicy.producer_exception) {
-                            commandSender.sendErrorResponse(requestId,
-                                    ServerError.ProducerBlockedQuotaExceededException,
-                                    illegalStateException.getMessage());
-                        }
-                        producerFuture.completeExceptionally(illegalStateException);
-                        producers.remove(producerId, producerFuture);
-                        return;
+                CompletableFuture<Void> backlogQuotaCheckFuture = CompletableFuture.allOf(
+                        topic.checkBacklogQuotaExceeded(producerName, BacklogQuotaType.destination_storage),
+                        topic.checkBacklogQuotaExceeded(producerName, BacklogQuotaType.message_age));
+                backlogQuotaCheckFuture.exceptionally(throwable -> {
+                    //throwable should be CompletionException holding TopicBacklogQuotaExceededException
+                    BrokerServiceException.TopicBacklogQuotaExceededException exception =
+                            (BrokerServiceException.TopicBacklogQuotaExceededException) throwable.getCause();
+                    IllegalStateException illegalStateException = new IllegalStateException(exception);
+                    BacklogQuota.RetentionPolicy retentionPolicy = exception.getRetentionPolicy();
+                    if (retentionPolicy == BacklogQuota.RetentionPolicy.producer_request_hold) {
+                        commandSender.sendErrorResponse(requestId,
+                                ServerError.ProducerBlockedQuotaExceededError,
+                                illegalStateException.getMessage());
+                    } else if (retentionPolicy == BacklogQuota.RetentionPolicy.producer_exception) {
+                        commandSender.sendErrorResponse(requestId,
+                                ServerError.ProducerBlockedQuotaExceededException,
+                                illegalStateException.getMessage());
                     }
-                }
-                // Check whether the producer will publish encrypted messages or not
-                if ((topic.isEncryptionRequired() || encryptionRequireOnProducer) && !isEncrypted) {
-                    String msg = String.format("Encryption is required in %s", topicName);
-                    log.warn("[{}] {}", remoteAddress, msg);
-                    commandSender.sendErrorResponse(requestId, ServerError.MetadataError, msg);
+                    producerFuture.completeExceptionally(illegalStateException);
                     producers.remove(producerId, producerFuture);
-                    return;
-                }
+                    return null;
+                });
 
-                disableTcpNoDelayIfNeeded(topicName.toString(), producerName);
+                backlogQuotaCheckFuture.thenRun(() -> {
```

Comment:
Which thread will execute this code?

The current thread (in case we reach here and the futures are already completed) or some other thread on completion of the one of the two futures above.

I am not sure we have control over what's happening here

Reply:
The trigger thread is the same as the above `backlogQuotaCheckFuture.exceptionally`, it's current thread or ReadEntry call back thread.